audeer/core/io.py

```diff
@@ -474,11 +477,19 @@ def list_file_names(
 
     """
     path = safe_path(path)
-    if not os.path.isdir(path):
+
+    if os.path.isdir(path):
+        pattern = None
+    elif os.path.exists(path):
```

Comment:
If `path` exists and is not a directory it would be better to immediately return it (handling basenames first) and not call the helper function as this would need to first run `os.listdir()` which might be slow if we have a folder with millions of files.

Reply:
What should we do when `recursive` is to `True` in that case? E.g. if we have:

```
/root/file.txt
/root/sub/file.txt
```

and the user requests:

```
audeer.list_file_names('/root/file.txt`, recursive=True)
```

would you expect it returns only `/root/file.txt` or also `/root/sub/file.txt`. So far the behavior was to return also `/root/sub/file.txt`.

Reply:
I would expect that it should return only `'/root/file.txt'` if this matches a file as in this case.

Reply:
> a user wants to collect all paths with the same file name from all sub-folders

Yes, that's indeed a good use-case, but for that I would expect a requested path similar to `/root/*/file.txt`.
At least I don't understand why `'/root/file.txt'` should match a path that is `/root/sub/file.txt`.

Reply:
I guess you have to see the `file.txt` part as a sort of pattern just like with `file.*` but without wildcards. And if `/root/file.*` together with `recursive=True` returns all matching files from sub-folder, `file.txt` should do the same I think.

Reply:
To request `/root/*/file.txt` is of cause not possible at the moment, and I would also not support it.

One way to still get the result you are looking for is to force it to be a pattern instead of an actual existing file, e.g. `audeer.list_file_names("dir/*.txt", recursive=True)` or `audeer.list_file_names("dir/[f]ile.txt", recursive=True)` which is of cause also slightly strange ;)

Reply:
> you have to see the file.txt part as a sort of pattern just like with file.*

Maybe this is indeed the best option, and it is also the way it is implemented at the moment.
What we need to do then is to change the docstring from

```
path: path to file, directory or pattern
```
to something like
```
path: path to directory,
    or path to directory plus file matching pattern,
    e.g. ``'dir/file.txt'``.
    If ``recursive`` is ``True`` this would return all files named ``file.txt``
    from all sub-folders.
    Besides the filename `*`, `?`, `[seq]`, and `[!seq]` can be used
    as explained in :mod:`fnmatch`.
```

Reply:
> To request /root/*/file.txt is of cause not possible at the moment, and I would also not support it.

Right, that's what the `recursive` argument is for.

> which is of cause also slightly strange ;)

Yes, that's why I suggest we simply stick with the current behavior.

> audeer.list_file_names("dir/[f]ile.txt", recursive=True)

Hehe, good you mentioned that `[...]` is also a valid pattern. We currently do not recognize it as an pattern. Will change that.

Reply:
I changed to also recognize `[...]` as a pattern and extended the tests and docstring examples accordingly. To make it easier to write docstring examples the working directory is now set to a temporal directory, I.e. we no longer need to manually clean up temporal files afterwards.

Reply:
I also changed the docstring of `path` to:

![image](https://user-images.githubusercontent.com/10383417/216096591-e40ae9fe-c0c1-4918-920c-f29f205f2e4f.png)